iree/compiler/Codegen/LLVMCPU/LLVMCPUTileFuseAndVectorizeLinalgTensorOps.cpp

```diff
@@ -139,71 +147,108 @@ void LLVMCPUTileFuseAndVectorizePass::runOnOperation() {
+    if (shouldTileReductionLoop) {
+      tileReductionPatterns.insert<TileWorkgroups>(
+          context,
+          linalg::LinalgTilingOptions().setTileSizeComputationFunction(
+              [](OpBuilder &builder,
+                 Operation *operation) -> SmallVector<Value, 4> {
+                auto tiles =
+                    getTileSizes(builder, operation,
+                                 static_cast<unsigned>(TilingLevel::L1Tiles));
+                auto numParallelLoops =
+                    dyn_cast<linalg::LinalgOp>(operation).getNumParallelLoops();
+                auto zeroTileVal = builder.create<arith::ConstantIndexOp>(
+                    operation->getLoc(), 0);
+                SmallVector<Value> reductionTiles(tiles.size(), zeroTileVal);
+                for (int i = numParallelLoops; i < tiles.size(); ++i) {
+                  reductionTiles[i] = tiles[i];
+                }
+                return std::move(reductionTiles);
+              }),
+          linalg::LinalgTransformationFilter(
+              ArrayRef<Identifier>{},
+              Identifier::get(getVectorizeMarker(), context)));
 
-    if (failed(applyPatternsAndFoldGreedily(
-            funcOp, std::move(tileReductionPatterns)))) {
-      return signalPassFailure();
+      if (failed(applyPatternsAndFoldGreedily(
+              funcOp, std::move(tileReductionPatterns)))) {
+        return signalPassFailure();
+      }
+      // Apply canoncalization
+      if (failed(applyTileAndFuseCanonicalizationPatterns(funcOp))) {
+        return signalPassFailure();
+      }
+      DEBUG_WITH_TYPE(DEBUG_TYPE, {
+        llvm::dbgs() << "\n--- After tiling reduction loops ---\n";
+        funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+        llvm::dbgs() << "\n\n";
+      });
     }
-    // Apply canoncalization
-    if (failed(applyTileAndFuseCanonicalizationPatterns(funcOp))) {
-      return signalPassFailure();
+  }
+
+  funcOp.walk([&](linalg::ContractionOpInterface op) {
+    if (failed(linalg::vectorizeLinalgOpPrecondition(op))) {
+      lowerToVectors = false;
     }
-    DEBUG_WITH_TYPE(DEBUG_TYPE, {
-      llvm::dbgs() << "\n--- After tiling reduction loops ---\n";
-      funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
-      llvm::dbgs() << "\n\n";
-    });
+  });
+  if (!lowerToVectors) {
+    // Apply second level of tiling patterns if they are not vectorizable. This
+    // will trigger LLVM auto-vectorization, which gains better performance.
+    {
+      funcOp.walk([&](linalg::ContractionOpInterface op) {
+        setMarker(op, getWorkgroupL1TileMarker());
+      });
+      OwningRewritePatternList l2patterns(&getContext());
+      l2patterns.insert<TileWorkgroups>(
+          context,
+          linalg::LinalgTilingOptions().setTileSizeComputationFunction(
+              [](OpBuilder &builder, Operation *op) -> SmallVector<Value, 4> {
+                return getTileSizes(
```

Comment:
IIUC this is level of tiling is same as the `native_vector_size`. I would drop the specification of this in `tile_sizes` and just use the values from `native_vector_size` here.

Reply:
Had an offline conversation with Hanhan. No need to address this right now. Once we move to use codegen driver, will revisit what this should look like.